lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,300 @@
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      private
+
+      def add_clerk(post, options)
+        post[:clerk] = {}
+        post[:clerk][:numericId] = options[:numeric_id]
+      end
+
+      def add_invoice(post, money, options)
+        post[:amount] = {}
+        post[:amount][:total] = money
+        post[:amount][:tax] = options[:amount][:tax] || 0.0
+      end
+
+      def api_options(post, options)
+        post[:apiOptions] = options[:api_options] if options[:api_options].present?
+      end
+
+      def add_credentials(post, options)
+        post[:dateTime] = options[:date_time] || DateTime.now.strftime('%Y-%m-%dT%H:%M:%S.%N+%H:%M')
+        post[:credential] = {}
+        post[:credential][:clientGuid] = options[:credential][:client_guid]
+        post[:credential][:authToken] = options[:credential][:auth_token]
+      end
+
+      def add_datetime(post, options)
+        post[:dateTime] = options[:dateTime] || DateTime.now.strftime('%Y-%m-%dT%H:%M:%S.%N+%H:%M')
+      end
+
+      def add_transaction(post, options)
+        post[:transaction] = {}
+        post[:transaction][:invoice] = options[:invoice]
+        post[:transaction][:notes] = options[:notes] if options[:notes].present?
+        if options[:purchase_card].present?
+          post[:transaction][:purchaseCard] = {}
+          post[:transaction][:purchaseCard][:customerReference] = options[:purchase_card][:customer_reference]
+          post[:transaction][:purchaseCard][:destinationPostalCode] = options[:purchase_card][:destination_postal_code]
+          post[:transaction][:purchaseCard][:productDescriptors] = options[:purchase_card][:product_descriptors]
+        end
+        post[:overrideBusDate] = options[:override_bus_date] if options[:override_bus_date].present?
+        if options[:amex].present?
+          post[:amex] = {}
+          post[:amex][:propertyCode] = options[:amex][:property_code]
+        end
+        if options[:auto].present?
+          post[:auto] = {}
+          post[:auto][:estimatedDays] = options[:auto][:estimated_days]
+        end
+        if options[:card_on_file].present?
+          post[:cardOnFile] = {}
+          post[:cardOnFile][:indicator] = options[:card_on_file][:indicator]
+          post[:cardOnFile][:usageIndicator] = options[:card_on_file][:usage_indicator]
+          post[:cardOnFile][:scheduledIndicator] = options[:card_on_file][:scheduled_indicator]
+          post[:cardOnFile][:indicator] = options[:card_on_file][:transaction_id]
+        end
+        if options[:hotel].present?
+          post[:hotel] = {}
+          post[:hotel][:estimatedDays] = options[:hotel][:estimated_days]
+        end
+        if options[:pin].present?
+          post[:pin] = {}
+          post[:pin][:block] = options[:pin][:block]
+          post[:pin][:ksn] = options[:pin][:ksn]
+        end
+      end
```

Comment:
This method looks a bit large and it's not clear that all of the values are related? I would expect shared values added to the `post` body to be sub options for the primary hash field (in this case, `transaction`)

Can we put each of the conditionals (`overrideBusDate`, `amex`, `auto`, `card_on_file`, `hotel`, `pin`) into their own helper methods and then we can call those methods from here?